rpcs3/rpcs3qt/user_manager_dialog.cpp

```diff
@@ -0,0 +1,373 @@
+	char* buf;
+	sprintf(buf, "%08d", nextLargest);
+	std::string nextUserId(buf);
+
+	// Create user folders and such.
+	const std::string homeDir = Emu.GetHddDir() + "home/";
+	const std::string userDir = homeDir + nextUserId;
+	fs::create_dir(homeDir);
+	fs::create_dir(fmt::format("%s/", userDir));
+	fs::create_dir(fmt::format("%s/exdata/", userDir));
+	fs::create_dir(fmt::format("%s/savedata/", userDir));
+	fs::create_dir(fmt::format("%s/trophy/", userDir));
+	fs::write_file(fmt::format("%s/localusername", userDir), fs::create + fs::excl + fs::write, username);
+}
+
+bool user_manager_dialog::ValidateUsername(const QString& textToValidate)
+{
+	// "Entire string (^...$) must be between 3 and 16 characters
+	// and only consist of letters, numbers, underscores, and hyphens."
+	QRegExpValidator validator(QRegExp("^[A-Za-z0-9_-]{3,16}$"));
+
+	int pos = 0;
+	QString text = textToValidate;
+	return (validator.validate(text, pos) == QValidator::Acceptable);
+}
+
+void user_manager_dialog::OnUserRename()
+{
+	QInputDialog* dialog = new QInputDialog(this);
+	dialog->setWindowTitle(tr("Rename User"));
+	dialog->setLabelText(tr("New Username: "));
+	dialog->resize(200, 100);
+
+	while (dialog->exec() != QDialog::Rejected)
+	{
+		dialog->resize(200, 100);
+		QString textToValidate = dialog->textValue();
+		if (!ValidateUsername(textToValidate))
+		{
+			QMessageBox::warning(this, tr("Error"), tr("Name must be between 3 and 16 characters and only consist of letters, numbers, underscores, and hyphens."));
+			continue;
+		}
+		int idx = m_table->currentRow();
+		int idx_real = m_table->item(idx, 0)->data(Qt::UserRole).toInt();
+
+		const std::string userId = m_user_list[idx_real]->GetUserId();
+		const std::string usernameFile = Emu.GetHddDir() + "home/" + userId + "/localusername";
+		fs::write_file(usernameFile, fs::rewrite, textToValidate.toStdString());
+		UpdateTable();
+		break;
+	}
+}
+
+void user_manager_dialog::OnUserCreate()
+{
+	QInputDialog* dialog = new QInputDialog(this);
+	dialog->setWindowTitle(tr("New User"));
+	dialog->setLabelText(tr("New Username: "));
+	dialog->resize(200, 100);
+
+	while (dialog->exec() != QDialog::Rejected)
+	{
+		dialog->resize(200, 100);
+		QString textToValidate = dialog->textValue();
+		if (!ValidateUsername(textToValidate))
+		{
+			QMessageBox::warning(this, tr("Error"), tr("Name must be between 3 and 16 characters and only consist of letters, numbers, underscores, and hyphens."));
+			continue;
+		}
+		GenerateUser(textToValidate.toStdString());
+		UpdateTable();
+		break;
+	}
+}
+
+void user_manager_dialog::OnUserLogin()
+{
+	int idx = m_table->currentRow();
+	int idx_real = m_table->item(idx, 0)->data(Qt::UserRole).toInt();
+	std::string selectedUserId = m_user_list[idx_real]->GetUserId();
+
+	m_selected_user = selectedUserId;
+	m_emu_settings->SetSetting(emu_settings::SelectedUser, m_selected_user);
+	m_emu_settings->SaveSettings();
+	UpdateTable();
```

Comment:
you can just disable the buttons here, no need for new fancy connects

Reply:
See that was my first thought, but I couldn't figure out how to retrieve the buttons from the dialog. I could make them member variables so I could access them from any function, but that felt like the "Get ok button" method, which I know you didn't like.

Reply:
of course you can either simply make them members or create another signal that gets emitted whenever you need to enable those buttons. eg:
``connect(this, &user_manager_dialog::OnUserLoginSuccess, enableButtons);``
and then use
``Q_EMIT OnUserLoginSuccess()``
on the appropriate place.

I don't really know what's the more appropriate way. members might be less confusing to maintain

Reply:
I'll opt for easier maintenance. But I didn't know you could emit signals on demand like that, pretty cool!

Reply:
Alright I lied - the Emit idea was definitely the easier to understand option.

Reply:
so is it all working now?

Reply:
I think so!